src/psyclone/transformations.py

```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+                                             nodes.OMPSingleDirective)
+                if valid:
+                    valid = valid and (not ancestor.nowait)
+                # If not valid, then we're in a Master Directive or a
+                # Single Directive with nowait. In this case we can't
+                # safely guarantee our forward dependency so throw an
+                # error
+                if not valid:
+                    raise TransformationError("Couldn't satisfy the "
+                                              "dependencies due to "
+                                              "taskloop dependencies "
+                                              "across barrierless OMP "
+                                              "serial regions.")
+
+    @staticmethod
+    def get_forward_dependence(taskloop, root):
+        '''
+        Returns the next forward dependence for a taskloop using the new
+        dependency analysis functions.
+        Forward dependencies can be of the following types:
+        Loop
+        OMPDoDirective
+        OMPTaskloopDirective
+        OMPTaskwaitDirective (If in same OMPSingle and that single has
+        nowait=False)
+        OMPSingleDirective (If ancestor OMPSingle has nowait=False)
+        OMPParallelDirective
+
+        The forward dependency is never a child of taskloop, and must have
+        abs_position > taskloop.abs_position
+
+        :param taskloop: the taskloop node to search for the
+                         forward_dependence.
+        :type taskloop: :py:class:`psyclone.psyir.nodes.OMPTaskloopDirective
+        :param root: the tree in which to search for the forward_dependence.
+        :type root: :py:class:`psyclone.f2pygen.DoGen`
+
+        :returns: the forward_dependence of taskloop.
+        :rtype: :py:class:`psyclone.f2pygen.DoGen`
+        '''
+        # We only look for specific types
+        node_list = root.walk((Loop, OMPDoDirective, OMPTaskloopDirective,
+                               OMPTaskwaitDirective, OMPSingleDirective,
+                               OMPParallelDirective))
+        taskloop_vars = VariablesAccessInfo()
+        for child in taskloop.walk(nodes.Node):
+            if child is not taskloop and not isinstance(child,
+                                                        (Schedule, Loop)):
+                taskloop_vars.merge(VariablesAccessInfo(child))
+        taskloop_signatures = taskloop_vars.all_signatures
+        # Find our parent serial region if it has a barrier
+        parent_single = taskloop.ancestor(OMPSingleDirective)
+        if parent_single is not None and parent_single.nowait:
+            parent_single = None
+        # Find our parent parallel region
+        parent_parallel = taskloop.ancestor(OMPParallelDirective)
+        for node in node_list:
+            if node.abs_position <= taskloop.abs_position:
+                continue
+            # Ignore our children
```

Comment:
s/our children/any descendents of the taskloop directive/

Reply:
Done